rpm/0035-Add-TrimmedSubject-message-sort-key.patch

```diff
@@ -0,0 +1,109 @@
+Date: Tue, 2 Nov 2021 20:55:47 +0300
+Subject: [PATCH] Add TrimmedSubject message sort key for sort by subject
+ without Re: and Fwd: prefixes
+
+---
+ .../qmfclient/qmailmessagesortkey.cpp         | 10 +++++++
+ src/libraries/qmfclient/qmailmessagesortkey.h |  4 ++-
+ src/libraries/qmfclient/qmailstore_p.cpp      | 26 ++++++++++++++++---
+ 3 files changed, 36 insertions(+), 4 deletions(-)
+
+diff --git a/src/libraries/qmfclient/qmailmessagesortkey.cpp b/src/libraries/qmfclient/qmailmessagesortkey.cpp
+index 661cdef6..ecf1c26d 100644
+--- a/src/libraries/qmfclient/qmailmessagesortkey.cpp
++++ b/src/libraries/qmfclient/qmailmessagesortkey.cpp
+@@ -272,6 +272,16 @@ QMailMessageSortKey QMailMessageSortKey::subject(Qt::SortOrder order)
+     return QMailMessageSortKey(Subject, order);
+ }
+ 
++/*!
++    Returns a key that sorts messages by their subject without Re:, Fw: and Fwd: prefixes, according to \a order.
++
++    \sa QMailMessage::trimmedSubject()
++*/
++QMailMessageSortKey QMailMessageSortKey::trimmedSubject(Qt::SortOrder order)
++{
++    return QMailMessageSortKey(TrimmedSubject, order);
++}
++
+ /*!
+     Returns a key that sorts messages by their origination timestamp, according to \a order.
+ 
+diff --git a/src/libraries/qmfclient/qmailmessagesortkey.h b/src/libraries/qmfclient/qmailmessagesortkey.h
+index cfe9b82d..5072d9f9 100644
+--- a/src/libraries/qmfclient/qmailmessagesortkey.h
++++ b/src/libraries/qmfclient/qmailmessagesortkey.h
+@@ -66,7 +66,8 @@ public:
+         ListId,
+         RestoreFolderId,
+         RfcId,
+-        ParentThreadId
++        ParentThreadId,
++        TrimmedSubject
+     };
+ 
+     typedef QMailSortKeyArgument<Property> ArgumentType;
+@@ -97,6 +98,7 @@ public:
+     static QMailMessageSortKey sender(Qt::SortOrder order = Qt::AscendingOrder);
+     static QMailMessageSortKey recipients(Qt::SortOrder order = Qt::AscendingOrder);
+     static QMailMessageSortKey subject(Qt::SortOrder order = Qt::AscendingOrder);
++    static QMailMessageSortKey trimmedSubject(Qt::SortOrder order = Qt::AscendingOrder);
+     static QMailMessageSortKey timeStamp(Qt::SortOrder order = Qt::AscendingOrder);
+     static QMailMessageSortKey receptionTimeStamp(Qt::SortOrder order = Qt::AscendingOrder);
+     static QMailMessageSortKey serverUid(Qt::SortOrder order = Qt::AscendingOrder);
+diff --git a/src/libraries/qmfclient/qmailstore_p.cpp b/src/libraries/qmfclient/qmailstore_p.cpp
+index 97f5020d..019ab269 100644
+--- a/src/libraries/qmfclient/qmailstore_p.cpp
++++ b/src/libraries/qmfclient/qmailstore_p.cpp
+@@ -632,6 +632,7 @@ static QMap<QMailMessageSortKey::Property, QMailMessageKey::Property> messageSor
+     map.insert(QMailMessageSortKey::Sender, QMailMessageKey::Sender);
+     map.insert(QMailMessageSortKey::Recipients, QMailMessageKey::Recipients);
+     map.insert(QMailMessageSortKey::Subject, QMailMessageKey::Subject);
++    map.insert(QMailMessageSortKey::TrimmedSubject, QMailMessageKey::TimeStamp);
```

Comment:
is Timestamp intentional?  Can TrimmedSubject be added to QMailMessageKey?

Reply:
>> is Timestamp intentional?
Yes - the order of the time is used at the SQL level when many equal subjects.

>>  Can TrimmedSubject be added to QMailMessageKey?
For what?

Reply:
Am I correct with the following: currently, your PR only supports using the trimmed subject for sorting; when performing a query from the database, the ORDER BY clause will use the Timestamp field in the query, and then the method which performs the query will then iterate over the (timestamp-sorted) results and then perform an additional sorting using the trimmed subject (via map insertion).

If so, there are a couple of potential issues:
1) specifying a DESC sort order will result in the timestamp ordering returned by the query being DESC order, but the additional sorting via map insertion will not be inverted.
2) sorting by the trimmed subject field will result in a potentially significant performance reduction, as the ordering cannot be done at the database query level, but is done in memory.

The existing QMF code basically uses that one mapping of sort key to mail message key to determine both the sort field and the filter field (at least, if I am reading the comment in messageSortMapInit() correctly), and for all other cases the mapping is defined such that the ORDER BY and SELECT fields are the same.  Your PR breaks this assumption, because it defines that the ORDER BY field is Timestamp but handles SELECT in code rather than via expansion from the mapping.

Also, the word "trimmed" already has a well-defined meaning in Qt code (i.e. trimmed of whitespace).  A better name might be `simplifiedSubject` or similar.

Conceptually, I would prefer if we can support this simplified subject more generally (that is, for both sorting and for filtering), without custom codepaths to handle this case specifically.  Presumably, it would require modifying QMailMessageHeader, adding the appropriate keys (which would most likely map to the normal Subject field, with a flag for subject simplification, and then adding a method somewhere like `QString simplifiedSubject(const QString &subj) const` which is used every time we query the subject for these cases - there are probably only a couple of places, e.g. when querying headers, and when querying messages).  Of course, this doesn't solve the performance hit, but that can be documented.

Reply:
>> but the additional sorting via map insertion will not be inverted.

Agree.

>> potentially significant performance reduction

SQLite also works in memory, so the performance hit shouldn't be too noticeable, but by means of sqlite we will not be able to carry out the sorting we need without creating additional fields in the database.

>> ...that is, for both sorting and for filtering...

I do not see any advantages of filtering by a simplified subject compared to a subject.

Reply:
> Also, the word "trimmed" already has a well-defined meaning in Qt code (i.e. trimmed of whitespace). A better name might be simplifiedSubject or similar.

May be originalSubject ?

Reply:
Regarding originalSubject - I agree, that sounds good to me.

Regarding the advantages of supporting filtering as well as sorting - the only real advantage is that it keeps the code consistent for all keys.  The current PR adds some cognitive load for future maintainers, because this specific key works differently to the rest.

Regarding the performance - ordinarily I would say: fair enough, better to avoid additional fields and complexity, so long as the performance overhead of current solution is very low or negligible.  However, maybe this is related to the above point about "consistent codepaths for all keys" - it might be cognitively simpler to add them.  I'm not sure.

@pvuorela @dcaliste I'd appreciate your thoughts here.  I am not familiar enough with QMF to have a hard stance on what is best; that said, the unique semantics for this key type does seem suboptimal to me, from a maintenance perspective.  A comprehensive commit message might help with that, I suppose.

Reply:
> Regarding originalSubject - I agree, that sounds good to me.

Changed.